dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactory.java

```diff
@@ -0,0 +1,358 @@
+     *    
+     * 3. An ApplicationStateListener. 
+     *    On startup the ASL only removes applications from the list of apps that have shut down so they will not return a no-op
+     *    On shutdown the ASL cleans up an OpenTelemetry object, removes its parent OpenTelemetryInfo from the cache, and registers it as an app which has shut down.
+     */
+
+    private static final String INSTRUMENTATION_NAME = "io.openliberty.microprofile.telemetry";
+    private static final String ENV_DISABLE_PROPERTY = "OTEL_SDK_DISABLED";
+    private static final String CONFIG_DISABLE_PROPERTY = "otel.sdk.disabled";
+    private static final String ENV_METRICS_EXPORTER_PROPERTY = "OTEL_METRICS_EXPORTER";
+    private static final String CONFIG_METRICS_EXPORTER_PROPERTY = "otel.metrics.exporter";
+    private static final String ENV_LOGS_EXPORTER_PROPERTY = "OTEL_LOGS_EXPORTER";
+    private static final String CONFIG_LOGS_EXPORTER_PROPERTY = "otel.logs.exporter";
+    private static final String SERVICE_NAME_PROPERTY = "otel.service.name";
+
+    private static final TraceComponent tc = Tr.register(OpenTelemetryInfoFactory.class);
+
+    private static final Map<J2EEName, OpenTelemetryInfo> appToInfo = new ConcurrentHashMap<J2EEName, OpenTelemetryInfo>();
+    private static final Set<J2EEName> shutDownApplications = Collections.newSetFromMap(new WeakHashMap<J2EEName, Boolean>()); //We have no good place to say forget about an app entirely, so I use weakreferences. 
+
+    //The key methods that provide access to telemetry objects begin here
+    public static OpenTelemetryInfo getOpenTelemetryInfo(J2EEName j2EEName) {
+        return appToInfo.computeIfAbsent(j2EEName, OpenTelemetryInfoFactory::createOpenTelemetryInfo);
+    }
+
+    public static Tracer getTracer(J2EEName j2EEName) {
+        try {
+            return getOpenTelemetryInfo(j2EEName).getOpenTelemetry().getTracer(INSTRUMENTATION_NAME);
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+            return OpenTelemetry.noop().getTracerProvider().get("");
+        }
+    }
+
+    public static Span getSpan() {
+        return new SpanProxy(); //TODO I cannot see a way to get no-op spans or baggage, check if we need code here to handle apps that have shut down.
+    }
+
+    public static Baggage getBaggage() {
+        return new BaggageProxy();
+    };
+
+    //The key methods that provide access to telemetry objects end here
+
+    private static OpenTelemetryInfo createOpenTelemetryInfo(J2EEName j2EEName) {
+        //If the app has already shut down but something still tries to get OpenTelemetry objects return a no-op object
+        if (shutDownApplications.contains(j2EEName)){ //There is a low risk race condition here. Thread A gets past this check just as thread B shuts everything down.
+            Tr.warning(tc, Tr.formatMessage(tc, "CWMOT5003.factory.used.after.shutdown"));
```

Comment:
Need to check message on this, should it be a general "internal problem warning"?

Also, I think it should contain the app name since we have it.

Reply:
Also, can we dump the stack trace in the trace log if this happens so we can help users diagnose why they saw this warning.